<a href="https://colab.research.google.com/github/NeznaikanaLune/MISIS_DS_Masters_degree_2020/blob/master/smriz_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**2. Содержательная постановка задачи**

Данную задачу можно сформулировать следующими способами:

**Вариант 1**
Определить оптимальный состав шихты для выплавки стали, нормированной по ГОСТу, с минимальной стоимостью материалов (используем минимальную стоимость выплавки единицы готовой продукции, полученную в предыдущей задаче) при условии, что выплавка ведется в 200 тонной печи и в наличие имеется ограниченное количество материалов.


**Вариант 2**
При наименьшей стоимости выплавки продукции в 200 тонной печи, удовлетворяющей ГОСТ определить максимально возможную загрузку.

**Вариант 3**
Определить максимальную загрузку печи при минимально возможной стоимости материалов(из всех вариантов с максимальной загрузкай выбрать вариант с наименьшей стоимостью) при условии, что выплавка ведется в 200 тонной печи и в наличие имеется ограниченное количество материалов.



# 3. **Формальная математическая модель**

## **Для варианта постановки задачи 1**
$X_1,X_2,X_3$ - количество тонн определенной шихты/смеси для выплавки стали

**Ограничения на количество материалов:**

$X_{1}+X_{2}+X_{3}\leq 200$

$X_{1}\leq 150$

$X_{2}\leq 100$

$X_{3}\leq 75$

**Дополнительные ограничения:**

$\sum(X) \geq 0$ - неотрицательный вес шихт

**Ограничения по ГОСТу:**

$\frac{(0.15X_1 + 0.15X_2 + 0.17X_3)}{X_{1}+X_{2}+X_{3}} \geq 0.16$

$\frac{(0.15X_100*0.15X_2 + 0.17X_3)}{X_{1}+X_{2}+X_{3}} \leq 0.18$

$\frac{(0.1X_1 + 0.08X_2 + 0.09X_3)}{X_{1}+X_{2}+X_{3}} \leq 0.09$

$\frac{(0.007X_1 + 0.003X_2 + 0.005X_3)}{X_{1}+X_{2}+X_{3}} \leq 0.005$

**Ограничение по стоимости стоимости:**

В данной задаче нужно максимально загрузить печь при минимальной стоимости шихт. За минимальную стоимость приму результат, полученный в задаче $2-это 150 \$$ за тонну. Но в данной задаче у нас есть $200$ тонная печь. А значит минимальная стоимость при максимальной загрузке такой печи можно условно оценикть, как $150*200=30000\$$. 

$(Price*x)<=30000$

**Целевая функция (максимизация загрузки):**

$\sum(X)\rightarrow \max$

## **Для варианта постановки задачи 2**

$X_1,X_2,X_3$ - доля определенной шихты/смеси для выплавки стали

**Ограничения по ГОСТу:**

$0.15x_1 + 0.15x_2 + 0.17x_3 \geq 0.16$

$0.15x_1 + 0.15x_2 + 0.17x_3 \leq 0.18$

$0.10x_1 + 0.08x_2 + 0.09x_3 \leq 0.09$

$0.007x_1 + 0.003x_2 + 0.005x_3 \leq 0.005$

$x \geq 0$

$\sum x= 1$ - используется хотя бы 1н исходный сплав

**На 1м этапе минимизируем целевую функцию стоимости с учетом ораничений выше:**

$(Price,x)\rightarrow \min$\
**На 2м этапе максимизируем загрузку печи с учетом полученных долей шихт при минимизации стоимости:**\
$\sum(Y)\rightarrow \max$\
**Ограничения только для 2го этапа:**\
$Y\geq0$\
$Y\leq200$\
$Y*X_{1}<=150$\
$Y*X_{2}<=100$\
$Y*X_{3}<=75$\
**Целевая функция:** $Y$-максимизация загрузки

## **Для варианта постановки задачи 3**
$X_1,X_2,X_3$ - количество тонн определенной шихты/смеси для выплавки стали

**Ограничения на количество материалов:**

$X_{1}+X_{2}+X_{3}\leq 200$

$X_{1}\leq 150$

$X_{2}\leq 100$

$X_{3}\leq 75$

**Дополнительные ограничения:**

$\sum(X) \geq 0$ - неотрицательный вес шихт

**Ограничения по ГОСТу:**

$\frac{(0.15X_1 + 0.15X_2 + 0.17X_3)}{X_{1}+X_{2}+X_{3}} \geq 0.16$

$\frac{(0.15X_100*0.15X_2 + 0.17X_3)}{X_{1}+X_{2}+X_{3}} \leq 0.18$

$\frac{(0.1X_1 + 0.08X_2 + 0.09X_3)}{X_{1}+X_{2}+X_{3}} \leq 0.09$

$\frac{(0.007X_1 + 0.003X_2 + 0.005X_3)}{X_{1}+X_{2}+X_{3}} \leq 0.005$

**Этап 1 находим максимально возможную суммарную загрузку печи с условием ограничений, описанных выше:**

Целевая функция для максимизации: $\sum(X)$

**Этап 2 находим минимальный вариант стоимости загрузки печи с учетом ограничения типа равенство на загрузку печи:**


Целевая функция для минимизации: $(Price,x)\rightarrow \min$\
Ограничение на суммарный вес материалов примет вид: $\sum(X)=$значение веса для оптимальной загрузки печи (На этапе 2 также нужно использовать все ограничения выше)




# 4. Алгоритм и ПО
• Python;

• Numpy - open-source модуль для python, который предоставляет общие математические и числовые операции в виде пре-скомпилированных, быстрых функций;

• cvxpy-python модуль для решения задач выпуклой оптимизации.

# 5. Решение задачи
Код можно запустить по [ссылке]([https://colab.research.google.com/drive/1doro8zipaTebYVSTj733srLGbIWBXq_C?usp=sharing). Это ссылка на работку, оформленную в Goggle Colab. Goggle Colab — это бесплатная среда для ноутбуков Jupyter, предоставляемая Google.

#### **Вариант_1**

In [ ]:
import numpy as np
import cvxpy
C = np.array([[0.15,0.10,0.007],[0.15,0.08,0.003],[0.17,0.09,0.005]]) #проценты записаны в долях
Price = np.array([300,200,150])#цена за тонну
Weight = np.array([[150,100,75]])#тонн доступно
x = cvxpy.Variable(shape=(1,C.shape[0]), integer = False)
total_cost = cvxpy.sum(x,axis=1)
C0 = x>=0
C1 = cvxpy.sum(x,axis=1)<=200
C2 = x*C[:,0] >=0.16*cvxpy.sum(x,axis=1)
C3 = x*C[:,0] <=0.18*cvxpy.sum(x,axis=1)
C4 = x*C[:,1] <=0.09*cvxpy.sum(x,axis=1)
C5 = x*C[:,2] <=0.005*cvxpy.sum(x,axis=1)
C6 = x[:,0]<= 150
C7 = x[:,1]<= 100
C8 = x[:,2]<= 75
C9 = x*Price<=200*150
problem = cvxpy.Problem(cvxpy.Maximize(total_cost), constraints=[C0,C1,C2,C3,C4,C5,C6,C7,C8,C9])
print('Загрузка печи:',problem.solve(),'\nСостав:',np.round(x.value,3),'\nСтоимость:',np.sum(x.value*Price))
print('Количество хрома в итоговой смеси:{} тонн,{} % от всей смеси'.format(round(np.sum(C[:,0]*x.value),2),round(np.sum(C[:,0]*x.value)/np.sum(x.value)*100,2)))
print('Количество никеля в итоговой смеси:{} тонн,{} % от всей смеси'.format(round(np.sum(C[:,1]*x.value),2),round(np.sum(C[:,1]*x.value)/np.sum(x.value)*100,2)))
print('Количество фосфора в итоговой смеси:{} тонн,{} % от всей смеси'.format(round(np.sum(C[:,2]*x.value),2),round(np.sum(C[:,2]*x.value)/np.sum(x.value)*100,2)))

Загрузка печи: 149.99999999782153 
Состав: [[21.574 53.426 75.   ]] 
Стоимость: 28407.38086929057
Количество хрома в итоговой смеси:24.0 тонн,16.0 % от всей смеси
Количество никеля в итоговой смеси:13.18 тонн,8.79 % от всей смеси
Количество фосфора в итоговой смеси:0.69 тонн,0.46 % от всей смеси


#### **Вариант_2**

In [ ]:
import numpy as np
import cvxpy

C = np.array([[0.15,0.10,0.007],[0.15,0.08,0.003],[0.17,0.09,0.005]])
Weight = np.array([[150,100,75]])#тонн доступно

Price = np.array([[300],[200],[150]])

x = cvxpy.Variable(shape=(1,C.shape[0]), integer = False)
total_cost = x*Price
C0 = x>=0
C1 = cvxpy.sum(x,axis=1)==1
C2 = x*C[:,0] >=0.16
C3 = x*C[:,0] <=0.18
C4 = x*C[:,1]<=0.09
C5 = x*C[:,2]<=0.005

problem = cvxpy.Problem(cvxpy.Minimize(total_cost), constraints=[C0,C1,C2,C3,C4,C5],)
print('Стоимость единичной выплавки:',problem.solve(),'\nДоли шихт:',np.round(x.value))


y = cvxpy.Variable(shape=1, integer = False)
x = np.round(x.value[0])
Weight = np.array([150,100,75])
total_weight = y
C0 = y>=1
С1 = (y*x <= Weight)
C2 = y<=200
problem = cvxpy.Problem(cvxpy.Maximize(total_weight), constraints=[C0,С1,C2])
print('Загрузка печи:',round(problem.solve(),2),'тонн')
print('Стоимость загрузки:',round(np.sum(y.value*x*Price.reshape(1,-1)),2),'$')

Стоимость единичной выплавки: 149.99999999502313 
Доли шихт: [[-0. -0.  1.]]
Загрузка печи: 75.0 тонн
Стоимость загрузки: 11250.0 $


#### **Вариант_3**

In [ ]:
import numpy as np
import cvxpy

C = np.array([[0.15,0.10,0.007],[0.15,0.08,0.003],[0.17,0.09,0.005]])

Weight = np.array([150,100,75])#тонн доступно
Price = np.array([300,200,150])

x = cvxpy.Variable(shape=(1,C.shape[0]), integer = False)
total_weight = cvxpy.sum(x)
C0 = x>=0
C1 = cvxpy.sum(x,axis=1)<=200
C2 = x*C[:,0] >=0.16*cvxpy.sum(x,axis=1)
C3 = x*C[:,0] <=0.18*cvxpy.sum(x,axis=1)
C4 = x*C[:,1]<=0.09*cvxpy.sum(x,axis=1)
C5 = x*C[:,2]<=0.005*cvxpy.sum(x,axis=1)
C6 = x[:,0]<= 150
C7 = x[:,1]<= 100
C8 = x[:,2]<= 75

problem = cvxpy.Problem(cvxpy.Maximize(total_weight), constraints=[C0,C1,C2,C3,C4,C5,C6,C7,C8])
print('Максимальный вес загрузки:',problem.solve())

total_cost = x*Price
C1 = cvxpy.sum(x,axis=1)==150 # Используем суммарный максимальновозможный вес загрузки, как ограничение типа равенство


problem = cvxpy.Problem(cvxpy.Minimize(total_cost), constraints=[C0,C1,C2,C3,C4,C5,C6,C7,C8],)
print('Доли каждой шихты:',problem.solve(),'\nСостав:',np.round(x.value))

print('Количество хрома в итоговой смеси:{} тонн,{} % от всей смеси'.format(round(np.sum(C[:,0]*x.value),2),round(np.sum(C[:,0]*x.value)/np.sum(x.value)*100,2)))
print('Количество никеля в итоговой смеси:{} тонн,{} % от всей смеси'.format(round(np.sum(C[:,1]*x.value),2),round(np.sum(C[:,1]*x.value)/np.sum(x.value)*100,2)))
print('Количество фосфора в итоговой смеси:{} тонн,{} % от всей смеси'.format(round(np.sum(C[:,2]*x.value),2),round(np.sum(C[:,2]*x.value)/np.sum(x.value)*100,2)))

Максимальный вес загрузки: 149.99999998826215
Доли каждой шихты: 26249.99999784636 
Состав: [[-0. 75. 75.]]
Количество хрома в итоговой смеси:24.0 тонн,16.0 % от всей смеси
Количество никеля в итоговой смеси:12.75 тонн,8.5 % от всей смеси
Количество фосфора в итоговой смеси:0.6 тонн,0.4 % от всей смеси


# 6. Анализ полученных результатов
Было применено 3 подхода для решения задачи. Какое именно выбрать решение зависит от целей заказчика: важнее максимизировать загрузку или минизировать стоимость. 

Оптимальным в условиях такой неопределенности можно считать решение 3, при котором мы максимально загружаем печь на 150 тонн и из всех возможных вариантов стоимости загрузки выбираем наименьшей - 26250$. Постановка задачи по варианту 1н имеет целью также максимизировать загрузку при условном максимально возможном пороге цены. Однако подход примененнный для решения Варианта 1 дал больший по стоимости результат, чем для 3 постановки, при одинаковой величине загрузки - 150 тонн